fetchcode/package_registry/__init__.py

```diff
@@ -0,0 +1,150 @@
+
+from packageurl.contrib.route import NoRouteAvailable
+from packageurl import PackageURL
+from packageurl.contrib.route import Router
+import requests
+
+
+router = Router()
+
+
+@attrs
+class PackageURLData:
+    """
+    Represent the Data extracted from any URL/PURL
+    `homepage_url` : URL to the homepage for this package
+    `api_url` :  Return the package repository API URL to obtain structured data for this package
+    `vcs_url` : URL for VCS download
+    `codeview_url` : URL where the code can be browsed online
+    `bugtracking_url` : URL to the issue or bug tracker for this package
+    `download_url` :  Return the package repository download URL to download the actual archive of code of this package. 
+    `documentation_url` : URL where the documentation can be found for this package
+    `readme_url` : URL where readme can be found for this package
+    `reverse_dependencies_url` : URL where reverse dependencies can be found for this package
+    `author_url` : URL where authors can be found regarding this package
+    `tags` : List of tags for this package
+    `tags_download_urls` : List of download URLS for tags of this package
+    `tags_readme_urls` : List of readme URLS for tags of this package
+    `tags_dependency_urls` : List of dependency URLs for tags of this package 
+    """
+
+    homepage_url = attrib(default=None)
+    api_url = attrib(default=None)
+    vcs_url = attrib(default=None)
+    codeview_url = attrib(default=None)
+    bugtracking_url = attrib(default=None)
+    download_url = attrib(default=None)
+    documentation_url = attrib(default=None)
+    readme_url = attrib(default=None)
+    reverse_dependencies_url = attrib(default=None)
+    author_url = attrib(default=None)
+    tags = attrib(default=None)
+    tags_download_urls = attrib(default=None)
+    tags_readme_urls = attrib(default=None)
+    tags_dependency_urls = attrib(default=None)
+
+
+def info(url):
+    """
+    Return data according to the `url` string
+    `url` string can be purl too
+    """
+    if url:
+        try:
+            return router.process(url)
+        except NoRouteAvailable:
+            return
+
+
+def get_response(url):
+    """
+    Return json object for a `url` string
+    """
+    resp = requests.get(url)
+    if "json" in dir(resp):
+        return resp.json()
+
+    raise Exception("Response of this URL does not contain json object")
+
+
+@router.route("pkg:cargo/.*")
+def cargo_data(purl):
+    """
+    Return `data` mapping of various URLs, from the `purl` string
+    """
+    purl_data = PackageURL.from_string(purl)
+    base_path = "https://crates.io"
+    name = purl_data.name
+    version = purl_data.version
+    api_url = f"{base_path}/api/v1/crates/{name}"
+    download_url = f"{api_url}/{version}/download"
+    readme_url = f"{api_url}/{version}/readme"
+    response = get_response(api_url)
+    crate = response.get("crate", {})
+    versions = response.get("versions", [])
```

Comment:
I wonder if we want to make an http request every single time. It may be better to add an optional arg. This would also make testing easier as you wouldnt have to mock everything.

Reply:
I am a little bit confused, which lines are we talking about 99,100,101

Reply:
99 is where the http request is made, and 100 and 101 use this data. 

My main point is that someone might want to craft data from a purl without having to make an http request.

Reply:
Ohhkay, I am getting your point, but all data is not presented inside PURL like it only gives version and name for that package, we can only form readme and download URL from that info,  but info like tags or codeview_url is in the API, so we need to extract it from there, I am open to suggestions or approach on this.